polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -1211,31 +1257,98 @@ impl<T: Config> Pallet<T> {
 		ensure!(assets.len() <= MAX_ASSETS_FOR_TRANSFER, Error::<T>::TooManyAssets);
 		let value = (origin_location, assets.into_inner());
 		ensure!(T::XcmReserveTransferFilter::contains(&value), Error::<T>::Filtered);
-		let (origin_location, assets) = value;
-		let context = T::UniversalLocation::get();
-		let fees = assets
-			.get(fee_asset_item as usize)
-			.ok_or(Error::<T>::Empty)?
-			.clone()
-			.reanchored(&dest, context)
-			.map_err(|_| Error::<T>::CannotReanchor)?;
-		let max_assets = assets.len() as u32;
-		let assets: MultiAssets = assets.into();
-		let xcm = Xcm(vec![
-			BuyExecution { fees, weight_limit },
-			DepositAsset { assets: Wild(AllCounted(max_assets)), beneficiary },
-		]);
-		let mut message = Xcm(vec![
-			SetFeesMode { jit_withdraw: true },
-			TransferReserveAsset { assets, dest, xcm },
-		]);
-		let weight =
-			T::Weigher::weight(&mut message).map_err(|()| Error::<T>::UnweighableMessage)?;
-		let hash = message.using_encoded(sp_io::hashing::blake2_256);
-		let outcome =
-			T::XcmExecutor::execute_xcm_in_credit(origin_location, message, hash, weight, weight);
-		Self::deposit_event(Event::Attempted { outcome });
-		Ok(())
+		let (origin_location, mut assets) = value;
+
+		if fee_asset_item as usize >= assets.len() {
+			return Err(Error::<T>::Empty.into())
+		}
+		let mut fees = assets.swap_remove(fee_asset_item as usize);
+		let fees_transfer_type =
+			<T::XcmExecutor as AssetTransferSupport>::determine_for(&fees, &dest)
+				.map_err(Error::<T>::InvalidAsset)?;
+		let assets_transfer_type = if assets.is_empty() {
+			// Single asset to transfer (one used for fees where transfer type is determined above).
+			ensure!(fees_transfer_type != TransferType::Teleport, Error::<T>::Filtered);
+			fees_transfer_type
+		} else {
+			// Find reserve for non-fee assets.
+			Self::validate_assets_and_find_reserve(&assets, &dest)?
+		};
+
+		// Disallow (for now) different _remote_ reserves for assets and fees.
+		match (&fees_transfer_type, &assets_transfer_type) {
+			(TransferType::RemoteReserve(a), TransferType::RemoteReserve(b)) if a != b =>
+				return Err(Error::<T>::TooManyReserves.into()),
+			_ => (),
+		};
+
+		if fees_transfer_type == assets_transfer_type {
+			// Same reserve location (fees not teleportable), we can batch together fees and assets
+			// in same reserve-based-transfer.
+			assets.push(fees.clone());
+		} else {
+			// Different transfer types: we have to do separate transfers for fees and assets.
```

Comment:
This honestly sounds like a workaround on a leaky abstraction: there's nothing stopping you from first executing a `BurnAsset` locally for the assets that are about to be transferred, and then making a series of `ReserveAssetDeposited` and/or `ReceiveTeleportedAssets` instructions in the same XCM program.

I'm surmising the reason why you didn't do so was because you still wanted to leverage the XCM executor's flow for teleport assets and reserve asset transfers, which then made you send 2 different XCM programs, when 1 would've sufficed.

Reply:
If there is no remote/3rd chain involved for either fees or assets, then yes, single message could be done.

**But** when either fees or assets (not both) have reserve on remote chain, then sending 2 different XCM programs is inevitable.

And yes, under the constraint above, I tried leveraging XCM executor's existing flow for teleports/reserves to keep things as clear as possible :)

I'll add a commit on top for doing single XCMs where possible - and we can then choose whether we like it or not - current flow is less efficient but might be cleaner conceptually (and in code), let's see :)

Reply:
@KiChjang I've simplified everything based on https://github.com/paritytech/polkadot-sdk/pull/1672#issuecomment-1775033549 - but current version of PR still builds and sends two XCM programs to same destination when `fees` and `assets` use different mechanisms.

I agree we can rework the pallet to do it in a single XCM message, but experimenting with that resulted in much more messy code:
- you have to practically hardcode most combos in the `fees | assets | transfer-type * 3` matrix,
- recreate in `pallet-xcm` a non-negligible amount of logic from `xcm-executor`.

I am strongly in favor of code clarity and maintainability in this case vs efficiency (I don't believe 2 XCMs of size `x` and `y` should be much more expensive than single `x + y` sized message anyway).

Reply:
I managed to structure the logic in an acceptable way so as to transfer assets and fees with **single custom XCM**.

One (yet unaddressed) problem with it is that for fees teleports we'd need to expose `XcmExecutor::AssetTransactor` to `pallet-xcm` so it can do `AssetTransactor::check_out()` on assets locally burned for teleportation.

The single XCM approach does have the following UX advantage:
- allows user to add small amount of `fees` just enough to cover costs, even if lower than ED on destination and full transfer still works. With separate XCMs one for getting fees across followed by one to get assets across, user has to make sure their fees ED on destination is met (either through this xfer or prefunded or whatever).

Disadvantages are:
- more complexity in `pallet-xcm`
- leaking non-negligible amount of logic from `XcmExecutor` into `pallet-xcm`

@KiChjang @franciscoaguirre I leave the final decision up to you, but please let me know ASAP so I can continue work in the right direction + add required tests.